ci/github-script/bot: request reviewers

wolfgangwalther opened the pull request:
This migrates the bash code to request reviewers to github-script. This will allow multiple nice improvements later on, but at this stage it's mostly a reduction in code and complexity.

## Things done

- [x] Tested locally.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
ci/github-script/bot: request reviewers

This migrates the bash code to request reviewers to github-script. This
will allow multiple nice improvements later on, but at this stage it's
mostly a reduction in code and complexity.

wolfgangwalther committed:
ci/github-script/reviewers: improve "needs: reviewers" label

This should fix the bug where the "needs: reviewer" label was set too
early, just to be removed immediately, because reviewers were then
requested.

MattSturgeon commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,103 @@
+  )
+
+  const team_members = new Set(
+    (await Promise.all(Array.from(teams).map(getTeamMembers)))
```
We can simplify `Array.from(it).map(fn)` to `Array.from(it, fn)` in a few places

MattSturgeon commented on ci/github-script/bot.js:
```diff
@@ -373,6 +374,30 @@ module.exports = async ({ github, context, core, dry }) => {
+          // TODO: Create owner map similar to maintainer map.
+          owners: (await readFile(`${pull_number}/owners.txt`, 'utf-8')).split(
+            '\n',
+          ),
```
Oh, I see. They come from https://github.com/NixOS/nixpkgs/blob/a14bd6bcccc6de27d675ba97e35d9525068a8551/ci/github-script/bot.js#L330-L338

MattSturgeon reviewed (commented):
SGTM, I can't spot any issues reading through the diff.

wolfgangwalther committed:
ci/github-script/bot: disregard bot and ghost approvals

We technically counted bot approvals and approvals by deleted users for
the approval labels as well. The former don't exist, yet, but if they
were, I don't think we'd count them. The latter should arguably *not* be
counted, because we can't tell anymore *who* approved, so we can't put
any weight on it as reviewers.

This simplifies the logic, too.

wolfgangwalther commented:
Fixed a few things while testing, I'm thinking this should work now.

wolfgangwalther commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,125 @@
+          await github.rest.repos.checkCollaborator({
+            ...context.repo,
+            username,
+          })
```
One thing that could happen on this endpoint is that the nixpkgs-ci app doesn't have the right privileges to check it. It used to be the codeowners app doing that.

We *should* have most of these privileges in nixpkgs-ci from https://github.com/NixOS/nixpkgs/pull/457566, but I'd still rather have an org owner available when merging this just in case.

balsoft commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,129 @@
+    )
+    // false indicates, that we do have reviewers and don't need the "needs: reviewers" label.
+    return false
+  }
```
I know this is the old behavior, but maybe we should still request reviews in cases like this, maybe limiting to the "first" 15 reviewers (by some stable sorting method).

balsoft commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,129 @@
+    ...owners.filter((handle) => handle && !handle.includes('/')),
+  ])
+  log('reviewers - users', Array.from(users).join(', '))
+  console.log('users', Array.from(users))
```
This seems to duplicate the `log` on the line above

```suggestion
```

balsoft commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,129 @@
+      pull_number,
+      reviewers,
+    })
+  }
```
This is slightly different from the original script: if requesting a review from one user failed, the old script still requested reviews from all other users, whereas this will fail completely AFAIU.

See also your commit https://github.com/NixOS/nixpkgs/commit/034613f860fcd339bd2c20c8f6bc259a2f9dc034

balsoft replied:
Judging by the commit message in https://github.com/NixOS/nixpkgs/commit/034613f860fcd339bd2c20c8f6bc259a2f9dc034, I guess this is actually somewhat important, since we did this switch once already.

balsoft reviewed (commented):
Mostly LGTM. I found one nitpick, one possibility for a future change, and one behavior change from the old script.

balsoft commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,129 @@
+// Caching the list of team members saves API requests when running the bot on the schedule and
+// processing many PRs at once.
+const members = {}
```
Actually, I don't think this is used anywhere in the script.

Either I suggest removing it:

```suggestion
```

or make use of it for caching, as suggested by the comment

MattSturgeon replied:
I believe this was added in preparation for this TODO:

https://github.com/wolfgangwalther/nixpkgs/blob/fc69b69871461448b6549962ed9b2a7e5bd6fcd7/ci/github-script/bot.js#L389-L393

However, a better way to handle that may be for `bot.js` to pass `reviewers.js` a reference to its `getTeamMembers` function.

wolfgangwalther commented on ci/github-script/reviewers.js:
```diff
@@ -0,0 +1,128 @@
+          await github.rest.repos.checkCollaborator({
+            ...context.repo,
+            username,
+          })
```
I have a PR open to fix it in https://github.com/NixOS/nixpkgs/pull/459170, but obviously... the bot didn't ping anyone :D
